File: oasis-core/src/oasis/graphics/IndexBuffer.java
package oasis.graphics;

import java.nio.Buffer;
import java.nio.ShortBuffer;

import oasis.core.Oasis;
import oasis.graphics.internal.InternalBuffer;

public class IndexBuffer extends GraphicsResource<InternalBuffer> {

    private ShortBuffer buffer; 
    private int size;
    private BufferUsage usage; 
    
    private boolean needsUpdate = true; 
    
    public IndexBuffer(int indices, BufferUsage usage) {
        this.size = indices; 
        this.usage = usage; 
        this.buffer = Oasis.getDirectBufferAllocator().allocate(size * 2).asShortBuffer(); 
        Oasis.getGraphicsDevice().requestInternalIndexBuffer(this); 
    }
    
    public void upload() {
        if (needsUpdate) {
            internal.uploadBuffer(); 
            needsUpdate = false; 
        }
    }
    
    public void release() {
        super.release(); 
        needsUpdate = true; 
    }

    public Buffer getAndFlipBuffer() {
        buffer.position(getSizeInShorts()); 
        buffer.flip(); 
        return buffer; 
    }
    
    public BufferUsage getUsage() {
        return usage; 
    }

    public int getSizeInBytes() {
        return size * 2; 
    }
    
    public int getSizeInShorts() {
        return size; 
    }
    
    public int getIndexCount() {
        return size; 
    }

    public void resize(int indices) {
        if (indices > buffer.capacity()) {
            Oasis.getDirectBufferAllocator().free(buffer); 
            buffer = Oasis.getDirectBufferAllocator().allocate(indices * 2).asShortBuffer(); 
        }
        size = indices; 
        buffer.limit(size); 
    }

    public void getIndices(int start, int count, short[] out, int outOffset) {
        buffer.position(start); 
        buffer.get(out, outOffset, count); 
    }

    public void setIndices(int start, int count, short[] in, int inOffset) {
        buffer.position(start); 
        buffer.put(in, inOffset, count); 
        
        needsUpdate = true; 
    }
    
}

File: Oasis/Graphics/IndexBuffer.cpp
#include "Oasis/Graphics/IndexBuffer.h"

#include <string.h>

namespace Oasis
{

IndexBuffer::IndexBuffer(int startElements, BufferUsage usage)
    : HardwareBuffer(usage)
    , data()
{
    data.resize(startElements);
}

IndexBuffer::~IndexBuffer() {}

void IndexBuffer::GetData(int start, int numElements, short* out) const
{
    int s = start;
    int e = numElements * sizeof (short);

    memcpy(out, &data[s], e);
}

void IndexBuffer::SetElementCount(int numElements)
{
    data.resize(numElements);

    UploadGpuData(data.size() * sizeof (short), &data[0]);
}

void IndexBuffer::SetData(int start, int numElements, const short* in)
{
    for (int i = 0; i < numElements; i++)
    {
        data[start + i] = in ? in[i] : 0;
    }

    UploadGpuSubData(start * sizeof (short), numElements * sizeof (short), in ? in : NULL);
}

}
